2-collectors/scala-stream-collector/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -255,19 +287,73 @@ class KinesisSink private (
                   s"Record failed with error code [${f._2.getErrorCode}] and message [${f._2.getErrorMessage}]"
                 )
             )
-            log.error(s"Retrying all failed records in $nextBackoff milliseconds...")
-            val failures = failurePairs.map(_._1)
-            scheduleBatch(failures, nextBackoff)
+            maybeSqs match {
+              case Some(sqs) =>
+                log.info(s"Sending all failed records to SQS buffer queue: ${sqs.sqsBufferName}")
+                putToSqs(sqs, failurePairs.map(_._1))
+              case None =>
+                log.error(
+                  s"Dropping ${failurePairs.size} failed events (consider setting a SQS Buffer in config.hocon)"
+                )
+            }
           }
         }
         case Failure(f) => {
           log.error("Writing failed.", f)
-          log.error(s"Retrying in $nextBackoff milliseconds...")
-          scheduleBatch(batch, nextBackoff)
+          maybeSqs match {
+            case Some(sqs) =>
+              log.info(
+                s"Sending all (${batch.size}) events from a batch to SQS buffer queue: ${sqs.sqsBufferName}"
+              )
+              putToSqs(sqs, batch)
+            case None =>
+              log.error(
+                s"Dropping all (${batch.size}) events from a batch (consider setting a SQS Buffer in config.hocon)"
+              )
+          }
         }
       }
     }
 
+  private def putToSqs(sqs: SqsClientAndName, batch: List[(ByteBuffer, String)]): Future[Unit] =
+    Future {
+      log.info(s"Writing ${batch.size} messages to SQS queue: ${sqs.sqsBufferName}")
+      val encoded = batch.map {
+        case (msg, key) =>
+          // The key is either IP or random UUID, so neither should contin `|`.
+          // Sqs doesn't support keys, so `|` is used as a separator. The same needs to be set up in
+          // sqs2kinesis project to decode the key - value pair.
+          val msgWithKey =
+            ByteBuffer.wrap(
+              Array.concat(key.getBytes, Array(sqsKeyValueSeparator.toByte), msg.array)
+            )
+          val b64Encoded = encode(msgWithKey)
+          new SendMessageBatchRequestEntry(UUID.randomUUID.toString, b64Encoded)
+      }
+      val MaxSqsBatchSize = 10
+      encoded.grouped(MaxSqsBatchSize).foreach { encodedGroup =>
+        val batchRequest =
+          new SendMessageBatchRequest()
+            .withQueueUrl(sqs.sqsBufferName)
+            .withEntries(encodedGroup.asJava)
+        Try {
+          sqs.sqsClient.sendMessageBatch(batchRequest)
+          log.info(
+            s"Batch of ${encodedGroup.size} was sent to SQS queue: ${sqs.sqsBufferName}"
+          )
+          ()
+        }.recover {
+          case e =>
+            log.error(s"Error sending to SQS queue(${sqs.sqsBufferName}): ${e.getMessage()}")
+        }
+      }
+    }
+
+  private def encode(bufMsg: ByteBuffer): String = {
+    val buffer = java.util.Base64.getEncoder.encode(bufMsg)
```

Comment:
This is a common practice to use underlying JVM locale for encoding and in the long run we probably should standardize upon a single charset. @oguzhanunlu has seen some misbehaviour caused by the kind of behavior.
Not expecting to do anything about it now, but to flag the issue for all to know.

Reply:
Just to share my particular case, running old version of event recovery resulted in successfully recovered events where some of the events had fields containing `?` instead of expected letters, whereas other recovered events had expected letters. It was because our software relied on system default and we've run it across dataflow runners, where some runners couldn't encode letters like `é`.

I don't expect anything on this without team consensus on charset discussion but we should be aware of the possibility esp. when we have limited or no control over deployment platform.

Reply:
Thanks, good point! What do you think about setting `utf-8` here and where decode happens in sqs2kinesis?

Reply:
> about setting utf-8 here

I think we should stay as is w.r.t. charset and have the ability to configure charset if we don't have charset control over the deployment platform. (e.g. dataflow)